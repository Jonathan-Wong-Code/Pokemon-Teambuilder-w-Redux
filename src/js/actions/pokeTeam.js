import { ADD_TEAM, DELETE_TEAM, EDIT_TEAM, SET_SAVED_TEAMS } from './types';
import database from './../firebase/firebase';
import regeneratorRuntime from "regenerator-runtime";

export const addPokeTeam = (pokeTeam) => async (dispatch, getState) => {
  const uid = getState().auth.uid;
  const response = await database.ref(`users/${uid}/pokemon`).push(pokeTeam);
  const newTeamObject = ({
    ...pokeTeam,
    id: response.key
  })

  dispatch({
    type : ADD_TEAM,
    team : newTeamObject
  });

  return response.key
};

export const deletePokeTeam = (id) => async (dispatch, getState) => {
  const uid = getState().auth.uid;
  database.ref(`users/${uid}/pokemon/${id}`).remove();

  dispatch({
    type : DELETE_TEAM,
    id
  });
};

export const editPokeTeam = (pokeTeam) => async (dispatch, getState) => {
  const uid = getState().auth.uid;
  database.ref(`users/${uid}/pokemon/${pokeTeam.id}`).update(pokeTeam);

  dispatch({
    type : EDIT_TEAM,
    team : pokeTeam
  });
};

export const setSavedTeams = () => async (dispatch, getState) => {
  const uid = getState().auth.uid;
  const response = await database.ref(`users/${uid}/pokemon`).once('value');
  const savedTeams = {};
  response.forEach(team => {
    savedTeams[team.key] = {
      ...team.val(),
      id : team.key
    }
  });

  dispatch({
    type: SET_SAVED_TEAMS,
    savedTeams
  });
}
