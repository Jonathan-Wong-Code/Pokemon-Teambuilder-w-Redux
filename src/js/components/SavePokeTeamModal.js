import React from 'react';

class SavePokeTeamModal extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      name : this.props.pokeTeam ? this.props.pokeTeam.name : '',
      description : this.props.pokeTeam ? this.props.pokeTeam.description : '',
      error :''
    };
  }

  handleInputChange = (e) => {
    this.setState({ name : e.target.value });
  }

  handleTextAreaChange = (e) => {
    this.setState({ description : e.target.value });
  }

  onBackClick = () => {
    this.props.toggleModal();
  }

  handleSubmit = (e) => {
    e.preventDefault();
    if(!this.state.name) {
      this.setState({ error : 'Must enter a team name to save.'});
    } else {
      this.props.handleSaveTeam(this.state.name, this.state.description);
      this.props.history.push('/SavedTeams')
    }
  }

  renderMessage = () => {
    return (
      this.state.error ? `${this.state.error}`: `Enter a name and description for your team.`
    )
  }

  render() {
    return (
      <div className='modal__background' 
        onClick={() => this.props.toggleModal()}
        data-test='save-form-component'
      >
        <div className='save-form modal__body' 
          onClick={(e) => e.stopPropagation()}
          data-test='save-form-modal-body'
        >
          <form action='' className='save-form__form' 
            onSubmit={this.handleSubmit}
            data-test='save-form-form'
          >
            <h2 className='save-form__message' data-test='save-form-message'>
              {this.renderMessage()}
            </h2>

            <label htmlFor='team-name' className='visuallyhidden'>Enter Team Name</label>
            <input 
              type='text' 
              placeholder='Team Name' 
              value={this.state.name}
              onChange={this.handleInputChange}
              className='save-form__name save-form__input'
              id='team-name'
              data-test='save-form-input'
            />

            <label 
              htmlFor='team-description' 
              className='visuallyhidden'>
              Enter Team Description
            </label>

            <textarea 
              name='' 
              id='' 
              cols='20' 
              rows='10'
              placeholder='Team Description' 
              value={this.state.description}
              onChange={this.handleTextAreaChange}
              className='save-form__description save-form__input'
              id='team-description'
              data-test='save-form-textarea'
            />

            <div className='save-form__buttons'>
              <button 
                className='save-form__save save-form__button btn' 
                type='submit'
                data-test='save-form-submit'
              >
                {this.props.type ==='create' ? 'Save Team' : 'Edit Team'}
              </button>
              
              <button 
                className='save-form__back save-form__button btn' 
                onClick={this.onBackClick}
                type='button'
                data-test='save-form-back'
              >
              Back 
              </button>
            </div>
          </form>
        </div>
      </div>  
    );
  }
};

export default SavePokeTeamModal;
