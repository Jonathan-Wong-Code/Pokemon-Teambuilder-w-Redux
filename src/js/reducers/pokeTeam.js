import { ADD_TEAM, DELETE_TEAM, EDIT_TEAM, SET_SAVED_TEAMS, LOGOUT } from './../actions/types';

const pokeTeamReducer = (state = {}, action) => {
  switch(action.type) {

    case ADD_TEAM :
    return {...state, [action.team.id] : action.team }

    case DELETE_TEAM :
    const newState = Object.assign({}, state);
    delete newState[action.id];
    return newState;

    case EDIT_TEAM :
    return {...state, [action.team.id] : action.team}

    case SET_SAVED_TEAMS:
    return action.savedTeams;

    case LOGOUT:
    return {}

    default:
    return state;
  }
}

export default pokeTeamReducer;
