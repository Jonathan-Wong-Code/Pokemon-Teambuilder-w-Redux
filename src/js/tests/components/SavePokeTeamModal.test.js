import React from 'react';
import { mount } from 'enzyme';
import SavePokeTeamModal from '../../components/SavePokeTeamModal';
import { singlePokeTeam } from './../fixtures/pokeTeam';

describe('The SavePokeTeamModal component creating new team (No PokeTeam Props)', () => {
  let wrapper, toggleModal, handleSaveTeam, history, type;
  beforeEach(() => {
    handleSaveTeam = jest.fn();
    history = { push : jest.fn()};
    type = 'create';
    toggleModal = jest.fn();

    const defaultProps = {
      handleSaveTeam,
      history,
      type,
      toggleModal
    }
    wrapper = mount(<SavePokeTeamModal {...defaultProps} />)
  });

  it('Should render the SavePokeTeamModal', () => {
    expect(wrapper.find(`[data-test='save-form-component']`).length).toBe(1);
  });

  it('Should take a snapshot of the SavePokeTeam Modal', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Should close the modal when the modal background is clicked', () => {
    wrapper.find(`[data-test='save-form-component']`).simulate('click');
    expect(toggleModal).toHaveBeenCalled();
  });

  it('Should not close the modal when the modal body is clicked', () => {
    wrapper.find(`[data-test='save-form-modal-body']`).simulate('click');
    expect(toggleModal).not.toHaveBeenCalled();
  });

  it('Should have an input that the user can enter a team name into', () => {
    wrapper.find(`[data-test='save-form-input']`).simulate('change', {
      target : { value : 'test123' }
    });

    expect(wrapper.find(`[data-test='save-form-input']`).prop('value')).toBe('test123');
    expect(wrapper.state('name')).toBe('test123');
  });

  it('Should have a textarea the user can enter a team description into to', () => {
    wrapper.find(`[data-test='save-form-textarea']`).simulate('change', {
      target : { value : 'this is a test description' }
    });

    expect(wrapper.find(`[data-test='save-form-textarea']`).prop('value')).toBe('this is a test description');
    expect(wrapper.state('description')).toBe('this is a test description');
  });

  it('Should have a button with the text of Save Team', () => {
    expect(wrapper.find(`[data-test='save-form-submit']`).text()).toBe('Save Team');
  });

  it('Should close the modal when the back button is clicked', () => {
    wrapper.find(`[data-test='save-form-back']`).simulate('click');
    expect(toggleModal).toHaveBeenCalled();
  });

  it('Should successfully submit the form when a name and description are entered', () => {
    wrapper.find(`[data-test='save-form-input']`).simulate('change', {
      target : { value : 'test123' }
    });

    wrapper.find(`[data-test='save-form-textarea']`).simulate('change', {
      target : { value : 'this is a test description' }
    });

    wrapper.find(`[data-test='save-form-form']`).simulate('submit', {
      preventDefault : () => {}
    });

    expect(handleSaveTeam).toHaveBeenLastCalledWith('test123', 'this is a test description');
    expect(history.push).toHaveBeenLastCalledWith('/SavedTeams');
    expect(wrapper.state('error')).toBeFalsy();
    expect(wrapper.find(`[data-test='save-form-message']`).text()).toBe('Enter a name and description for your team.');
  });

  it('Should not submit the form when no team name is entered', () => {
    wrapper.find(`[data-test='save-form-input']`).simulate('change', {
      target : { value : '' }
    });

    wrapper.find(`[data-test='save-form-textarea']`).simulate('change', {
      target : { value : 'this is a test description' }
    });

    wrapper.find(`[data-test='save-form-form']`).simulate('submit', {
      preventDefault : () => {}
    });

    expect(handleSaveTeam).not.toHaveBeenCalled();
    expect(history.push).not.toHaveBeenCalled();
    expect(wrapper.state('error')).toBeTruthy();
    expect(wrapper.find(`[data-test='save-form-message']`).text()).toBe('Must enter a team name to save.');
  });
});

describe('The SavePokeTeamModal Component when team is being edited', () => {
  let wrapper, toggleModal, handleSaveTeam, history, type, pokeTeam;
  beforeEach(() => {
    handleSaveTeam = jest.fn();
    history = { push : jest.fn()};
    type = 'edit';
    toggleModal = jest.fn();
    pokeTeam = singlePokeTeam;

    const defaultProps = {
      handleSaveTeam,
      history,
      type,
      toggleModal,
      pokeTeam
    };
    wrapper = mount(<SavePokeTeamModal {...defaultProps} />);
  });

  it('Should render a snapshot of the component', () => {
    expect(wrapper).toMatchSnapshot();
  });
  
  it('Should render the component', () => {
    expect(wrapper.find(`[data-test='save-form-component']`).length).toBe(1);
  });

  it('Should set the name and description state when PokeTeam is being edited/provided as props', () => {
    expect(wrapper.state('name')).toBe(singlePokeTeam.name);
    expect(wrapper.state('description')).toBe(singlePokeTeam.description);
  });

  it('Should render the submit button with the correct text', () => {
    expect(wrapper.find(`[data-test='save-form-submit']`).text()).toBe('Edit Team');
  })
});
