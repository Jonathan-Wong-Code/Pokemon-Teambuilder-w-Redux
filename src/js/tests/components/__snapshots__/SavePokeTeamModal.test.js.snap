// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The SavePokeTeamModal Component when team is being edited Should render a snapshot of the component 1`] = `
<SavePokeTeamModal
  handleSaveTeam={[MockFunction]}
  history={
    Object {
      "push": [MockFunction],
    }
  }
  pokeTeam={
    Object {
      "description": "This is pokeTeam1",
      "id": "team1",
      "name": "pokeTeam1",
      "pokemon": Array [
        Object {
          "name": "pokemon1",
          "sprites": Object {
            "front_default": "sprite1",
          },
        },
        Object {
          "name": "pokemon2",
          "sprites": Object {
            "front_default": "sprite2",
          },
        },
        Object {
          "name": "pokemon3",
          "sprites": Object {
            "front_default": "sprite3",
          },
        },
        Object {
          "name": "pokemon4",
          "sprites": Object {
            "front_default": "sprite4",
          },
        },
        Object {
          "name": "pokemon5",
          "sprites": Object {
            "front_default": "sprite5",
          },
        },
        Object {
          "name": "pokemon6",
          "sprites": Object {
            "front_default": "sprite6",
          },
        },
      ],
    }
  }
  toggleModal={[MockFunction]}
  type="edit"
>
  <div
    className="modal__background"
    data-test="save-form-component"
    onClick={[Function]}
  >
    <div
      className="save-form modal__body"
      data-test="save-form-modal-body"
      onClick={[Function]}
    >
      <form
        action=""
        className="save-form__form"
        data-test="save-form-form"
        onSubmit={[Function]}
      >
        <h2
          className="save-form__message"
          data-test="save-form-message"
        >
          Enter a name and description for your team.
        </h2>
        <label
          className="visuallyhidden"
          htmlFor="team-name"
        >
          Enter Team Name
        </label>
        <input
          className="save-form__name save-form__input"
          data-test="save-form-input"
          id="team-name"
          onChange={[Function]}
          placeholder="Team Name"
          type="text"
          value="pokeTeam1"
        />
        <label
          className="visuallyhidden"
          htmlFor="team-description"
        >
          Enter Team Description
        </label>
        <textarea
          className="save-form__description save-form__input"
          cols="20"
          data-test="save-form-textarea"
          id="team-description"
          name=""
          onChange={[Function]}
          placeholder="Team Description"
          rows="10"
          value="This is pokeTeam1"
        />
        <div
          className="save-form__buttons"
        >
          <button
            className="save-form__save save-form__button btn"
            data-test="save-form-submit"
            type="submit"
          >
            Edit Team
          </button>
          <button
            className="save-form__back save-form__button btn"
            data-test="save-form-back"
            onClick={[Function]}
            type="button"
          >
            Back
          </button>
        </div>
      </form>
    </div>
  </div>
</SavePokeTeamModal>
`;

exports[`The SavePokeTeamModal component creating new team (No PokeTeam Props) Should take a snapshot of the SavePokeTeam Modal 1`] = `
<SavePokeTeamModal
  handleSaveTeam={[MockFunction]}
  history={
    Object {
      "push": [MockFunction],
    }
  }
  toggleModal={[MockFunction]}
  type="create"
>
  <div
    className="modal__background"
    data-test="save-form-component"
    onClick={[Function]}
  >
    <div
      className="save-form modal__body"
      data-test="save-form-modal-body"
      onClick={[Function]}
    >
      <form
        action=""
        className="save-form__form"
        data-test="save-form-form"
        onSubmit={[Function]}
      >
        <h2
          className="save-form__message"
          data-test="save-form-message"
        >
          Enter a name and description for your team.
        </h2>
        <label
          className="visuallyhidden"
          htmlFor="team-name"
        >
          Enter Team Name
        </label>
        <input
          className="save-form__name save-form__input"
          data-test="save-form-input"
          id="team-name"
          onChange={[Function]}
          placeholder="Team Name"
          type="text"
          value=""
        />
        <label
          className="visuallyhidden"
          htmlFor="team-description"
        >
          Enter Team Description
        </label>
        <textarea
          className="save-form__description save-form__input"
          cols="20"
          data-test="save-form-textarea"
          id="team-description"
          name=""
          onChange={[Function]}
          placeholder="Team Description"
          rows="10"
          value=""
        />
        <div
          className="save-form__buttons"
        >
          <button
            className="save-form__save save-form__button btn"
            data-test="save-form-submit"
            type="submit"
          >
            Save Team
          </button>
          <button
            className="save-form__back save-form__button btn"
            data-test="save-form-back"
            onClick={[Function]}
            type="button"
          >
            Back
          </button>
        </div>
      </form>
    </div>
  </div>
</SavePokeTeamModal>
`;
